### DISCUSSION TITLE
Docs: How to query translations using GraphQL?

### DISCUSSION THREAD

**Discussion by sapkra** [directus/directus#4487] [Archive]
I wanted to figure out how to get translations using GraphQL but it's not documented yet. Is it even possible?

https://docs.directus.io/reference/api/query/#deep

**Comment by rijkvanzanten** [comment]
A translation is no different from a regular o2m relationship, you can fetch. (I'm assuming you have a `articles` table with a `translations` here:

```graphql
query {
  items {
    articles {
      id
      translations {
        language_code
        title
      }
    }
  }
}
```

**Reply by sapkra** [reply]
But then I have to filter them again on the client. It would be much more performant to filter them on the database to only get the language I want to show. Isn't this possible yet?

**Reply by rijkvanzanten** [reply]
Yup, you can add the filter attribute on the `translations` layer 👍🏻 

```graphql
query {
  items {
    articles {
      id
      translations(filter: { language_code: { _eq: "en-US" }}) {
        language_code
        title
      }
    }
  }
}
```

See https://docs.directus.io/reference/api/query/#filter for more information

**Reply by sapkra** [reply]
Thank you. I'm new to directus and missed that. ^^

**Reply by maggie44** [reply]
Is this still the best way to filter by language? I have added a translated data model as per this video (which is really helpful): https://www.youtube.com/watch?v=HGQppTT23T8

I am then accessing the translated model through another model using the many-to many. 

My query now seems like is has to look like this:

```graphql
query resources {
  resources {
    license {
      id
      licenses_id {
        translations {
          license
          languages_id {
            code
          }
        }
      }
    }
  }
}

```

The above works but 'code' (the language code) doesn't exist on translations next to 'license', so it's not filterable as in your example. I can filter it from the next level down, but I'm curious if something has changed with the setup. The example given seems to make more sense in terms of structure of the data. 

Here is the output:

```json
{
    "data": {
        "resources": [
            {
                "license": []
            },
            {
                "license": [
                    {
                        "id": "1",
                        "licenses_id": {
                            "translations": [
                                {
                                    "license": "Free",
                                    "languages_id": {
                                        "code": "en-US"
                                    }
                                },
                                {
                                    "license": "Gratis",
                                    "languages_id": {
                                        "code": "es-ES"
                                    }
                                }
                            ]
                        }
                    }
                ]
            },
            {
                "license": []
            }
        ]
    }
} 
```

**Reply by rijkvanzanten** [reply]
@maggie0002 Would you be able to provide some more information on your schema? It's a little hard to tell from the above what we're looking at exactly.. I'm assuming "translations" is your translations field (that much is obvious 😁), but I'm not too sure what "resources" is in relation to the parent, or anything down to the "translations" itself. 

I _think_ what you're looking for is

```graphql
query resources {
  resources {
    license {
      id
      licenses_id {
        translations(filter: { languages_id: { code: { _eq: "nl-NL" } } }) {
          license
          languages_id {
            code
          }
        }
      }
    }
  }
}

```

(except you might wanna use `en-US` or `es-ES` instead of `nl-NL` 😉 🇪🇸 )

**Reply by maggie44** [reply]
`license` is a many-to-many with `resource`. And then the translations are setup on the `license` using the default settings.  

I did experiment with a schema similar to the one you provided there. It just seemed odd that they were on separate levels, whereas your example back in March brought them together, and wondered if I had set something up wrong.

In the end I discovered that the amount of translations I needed to do meant it was best not to use the Directus translation feature. Namely the number of translations and not being able to integrate with platforms like Weblate, and also not having a standard feature to resort to en-US if the field is missing for nl-NL. 

For others to reference if helpful, I added a `key` input field to each of my entries, then had a matching translation in my i18n file in the app that uses the same key. I then pulled the `key` through GraphQL and inserted it as the i18n variable to insert.
